torch/csrc/jit/passes/autocast.cpp

```diff
@@ -0,0 +1,456 @@
+        TORCH_INTERNAL_ASSERT(
+            !incompatible_amp.has_value() || incompatible_amp.value(),
+            "Calls are not expected with AMP & JIT");
+        incompatible_amp = true;
+        break;
+
+      case prim::CallMethod:
+        // TODO: limit it only to amp related node;
+        if (auto class_type = node->input(0)->type()->cast<ClassType>()) {
+          const auto& name = node->s(attr::name);
+          const auto& function = class_type->getMethod(name);
+          if (!function.isGraphFunction()) {
+            TORCH_INTERNAL_ASSERT(
+                !incompatible_amp.has_value() || incompatible_amp.value(),
+                "Calls are not expected with AMP & JIT");
+            incompatible_amp = true;
+          }
+        } else {
+          TORCH_INTERNAL_ASSERT(
+              !incompatible_amp.has_value() || incompatible_amp.value(),
+              "Unexpected prim::CallMethod form with AMP & JIT");
+          incompatible_amp = true;
+        }
+        break;
+
+      case prim::Enter:
+        if (auto autocast_scope =
+                parseAutocast(node->input(), current_state())) {
+          if (node->hasUses()) {
+            // TODO: better error message
+            AT_ERROR("`with autocast() as ...` is not supported");
+          }
+          TORCH_INTERNAL_ASSERT(
+              !incompatible_amp.has_value() || !incompatible_amp.value(),
+              "Unsupported case by AMP & JIT");
+          incompatible_amp = false;
+          autocast_stack.push(*autocast_scope);
+        }
+        break;
+
+      case prim::Exit:
+        if (isAutocastNode(node->input(0))) {
+          TORCH_INTERNAL_ASSERT(!autocast_stack.empty());
+          TORCH_INTERNAL_ASSERT(autocast_stack.top().instance == node->input());
+          TORCH_INTERNAL_ASSERT(
+              !incompatible_amp.has_value() || !incompatible_amp.value(),
+              "Unsupported case by AMP & JIT");
+          incompatible_amp = false;
+          autocast_stack.pop();
+        }
+        break;
+
+      // CastPolicy::fp16 (cast all inputs to float16)
+      case aten::_convolution:
+      case aten::_convolution_nogroup:
+      case aten::conv1d:
+      case aten::conv2d:
+      case aten::conv3d:
+      case aten::conv_tbc:
+      case aten::conv_transpose1d:
+      case aten::convolution:
+      case aten::cudnn_convolution:
+      case aten::cudnn_convolution_transpose:
+      case aten::prelu:
+      case aten::addmm:
+      case aten::addmv:
+      case aten::addr:
+      case aten::matmul:
+      case aten::mm:
+      case aten::mv:
+      case aten::linear:
+      case aten::addbmm:
+      case aten::baddbmm:
+      case aten::bmm:
+      case aten::chain_matmul:
+      case aten::_thnn_fused_lstm_cell:
+      case aten::_thnn_fused_gru_cell:
+      case aten::lstm_cell:
+      case aten::gru_cell:
+      case aten::rnn_tanh_cell:
+      case aten::rnn_relu_cell:
+        if (!node->schema().is_mutable()) {
```

Comment:
why would schema be mutable here

Reply:
I think the `is_mutable` is only to avoid inserting autocast before in-place operations.